File: tests/wpt/tests/scroll-to-text-fragment/stash.py
import time

def main(request, response):
    key = request.GET.first(b"key")

    if request.method == u"POST":
        # Received result data from target page
        request.server.stash.put(key, request.body, u'/scroll-to-text-fragment/')
        return u"ok"
    else:
        # Request for result data from test page
        value = request.server.stash.take(key, u'/scroll-to-text-fragment/')
        return value

File: tests/wpt/tests/scroll-to-text-fragment/stash.js
// Put test results into Stash
function stashResultsThenClose(key, results) {
  fetch(`/scroll-to-text-fragment/stash.py?key=${key}`, {
    method: 'POST',
    body: JSON.stringify(results)
  }).then(() => {
    window.close();
  });
}

// Fetch test results from the Stash
function fetchResults(key, resolve, reject) {
  fetch(`/scroll-to-text-fragment/stash.py?key=${key}`).then(response => {
    return response.text();
  }).then(text => {
    if (text) {
      try {
        const results = JSON.parse(text);
        resolve(results);
      } catch(e) {
        reject();
      }
    } else {
      // We keep trying to fetch results as the target page may not have stashed
      // them yet.
      fetchResults(key, resolve, reject);
    }
  });
}
